engine/src/main/java/org/pentaho/di/trans/StepWithMappingMeta.java

```diff
@@ -0,0 +1,354 @@
+    this.directoryPath = directoryPath;
+  }
+
+  @Override
+  public String[] getDirectories() {
+    return new String[]{ directoryPath };
+  }
+
+  @Override
+  public void setDirectories( String[] directories ) {
+    this.directoryPath = directories[0];
+  }
+
+  /**
+   * @return the fileName
+   */
+  public String getFileName() {
+    return fileName;
+  }
+
+  /**
+   * @param fileName the fileName to set
+   */
+  public void setFileName( String fileName ) {
+    this.fileName = fileName;
+  }
+  /**
+   * @param fileName the fileName to set
+   */
+  public void replaceFileName( String fileName ) {
+    this.fileName = fileName;
+  }
+
+  /**
+   * @return the transName
+   */
+  public String getTransName() {
+    return transName;
+  }
+
+  /**
+   * @param transName the transName to set
+   */
+  public void setTransName( String transName ) {
+    this.transName = transName;
+  }
+
+  /**
+   * @return the transObjectId
+   */
+  public ObjectId getTransObjectId() {
+    return transObjectId;
+  }
+
+  /**
+   * @param transObjectId the transObjectId to set
+   */
+  public void setTransObjectId( ObjectId transObjectId ) {
+    this.transObjectId = transObjectId;
+  }
+
+  @Override
+  public String exportResources( VariableSpace space, Map<String, ResourceDefinition> definitions,
+                                 ResourceNamingInterface resourceNamingInterface, Repository repository,
+                                 IMetaStore metaStore ) throws KettleException {
+    try {
+      // Try to load the transformation from repository or file.
+      // Modify this recursively too...
+      //
+      // NOTE: there is no need to clone this step because the caller is
+      // responsible for this.
+      //
+      // First load the mapping transformation...
+      //
+      TransMeta mappingTransMeta = loadMappingMeta( this, repository, metaStore, space );
+
+      // Also go down into the mapping transformation and export the files
+      // there. (mapping recursively down)
+      //
+      String proposedNewFilename =
+              mappingTransMeta.exportResources(
+                      mappingTransMeta, definitions, resourceNamingInterface, repository, metaStore );
+
+      // To get a relative path to it, we inject
+      // ${Internal.Entry.Current.Directory}
+      //
+      String newFilename = "${" + INTERNAL_VARIABLE_ENTRY_CURRENT_DIRECTORY + "}/" + proposedNewFilename;
+
+      // Set the correct filename inside the XML.
+      //
+      mappingTransMeta.setFilename( newFilename );
+
+      // exports always reside in the root directory, in case we want to turn
+      // this into a file repository...
+      //
+      mappingTransMeta.setRepositoryDirectory( new RepositoryDirectory() );
+
+      // change it in the job entry
+      //
+      replaceFileName( newFilename );
+
+      setSpecificationMethod( ObjectLocationSpecificationMethod.FILENAME );
+
+      return proposedNewFilename;
+    } catch ( Exception e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)